src/line_buffer.cc

```diff
@@ -116,9 +116,154 @@ class lock_hack {
 };
 /* XXX END */
 
+
+#define Z_BUFSIZE 65536U
+#define SYNCPOINT_SIZE (1024 * 1024)
+#define GZ_HEADER_MODE (15 + 32)    /*> Automatic zstd or gzip decoding */
+
+line_buffer::gz_indexed::gz_indexed()
+{
+    if ((this->inbuf = (Bytef *)malloc(Z_BUFSIZE)) == NULL) {
+        throw bad_alloc();
+    }
+}
+
+void line_buffer::gz_indexed::close()
+{
+    // Release old stream, if we were open
+    if (*this) {
+        inflateEnd(&this->strm);
+        ::close(this->gz_fd);
+        this->syncpoints.clear();
+        this->gz_fd = 0;
+    }
+}
+
+void line_buffer::gz_indexed::init_stream()
+{
+    if (*this) inflateEnd(&this->strm);
+
+    // initialize inflate struct
+    this->strm.zalloc = Z_NULL;
+    this->strm.zfree = Z_NULL;
+    this->strm.opaque = Z_NULL;
+    this->strm.avail_in = 0;
+    this->strm.next_in = Z_NULL;
+    this->strm.avail_out = 0;
+    int rc = inflateInit2(&strm, GZ_HEADER_MODE);
+    if (rc != Z_OK) {
+        throw(rc);  // FIXME: exception wrapper
+    }
+}
+
+void line_buffer::gz_indexed::open(int fd)
+{
+    this->close();
+    this->init_stream();
+    this->gz_fd = fd;
+}
+
+int line_buffer::gz_indexed::stream_data(void * buf, size_t size)
+{
+    this->strm.avail_out = size;
+    this->strm.next_out = (unsigned char *) buf;
+
+    int last = this->syncpoints.size() ?
+                    this->syncpoints.back().in : 0;
+    while (this->strm.avail_out) {
+        if (!this->strm.avail_in) {
+            int rc = ::pread(this->gz_fd,
+                        &this->inbuf[0],
+                        Z_BUFSIZE,
+                        this->strm.total_in);
+            if (rc < 0) {
+                return rc;
+            }
+            this->strm.next_in = this->inbuf;
+            this->strm.avail_in = rc;
+        }
+        if (this->strm.avail_in) {
+            int flush = last > this->strm.total_in
+                          ? Z_SYNC_FLUSH : Z_BLOCK;
+            auto err = inflate(&this->strm, flush);
+            if (err == Z_STREAM_END) {
+                log_info(" inflate: Z_STREAM_END");
+                break;
+            } else if (err != Z_OK) {
+                log_info(" inflate-error: %d", (int)err);
+                throw error(err);  // FIXME: exception wrapper
+            }
+
+            if (this->strm.total_in >= last + SYNCPOINT_SIZE &&
+                size > this->strm.avail_out + GZ_WINSIZE &&
+                (strm.data_type & 128) && !(strm.data_type & 64)) {
```

Comment:
nit: can you add some constants for the 128/64 magic numbers

Reply:
I tried to use `gzip.h` constants for these, but they seem to be nonexistent.  Disappointing.  So I used the code example from libz.  But I agree, these suck.  Will fix.